crash - markastemplate vsphere post-processor

- Packer Version = 1.1.2
- Host platform = CentOS Linux release 7.3.1611 (Core) / kernel = 3.10.0-514.6.2.el7.x86_64
- Debug log and template: https://gist.github.com/jbrowder2/0194bb6cea38cb742a0fab0900385774

I am dealing with issues with our environment. I unfortunately don't have the build environment containerized yet. We are using a VM for a jenkins slave currently. 

after git cloning and using go build I am getting:

checkpoint.go:8:2: cannot find package "github.com/hashicorp/go-checkpoint" in any of:
	/home/users/bhundven/projects/go/src/github.com/hashicorp/go-checkpoint (from $GOROOT)
	/root/go/src/github.com/hashicorp/go-checkpoint (from $GOPATH)
main.go:18:2: cannot find package "github.com/hashicorp/go-uuid" in any of:
	/home/users/bhundven/projects/go/src/github.com/hashicorp/go-uuid (from $GOROOT)
	/root/go/src/github.com/hashicorp/go-uuid (from $GOPATH)
checkpoint.go:9:2: cannot find package "github.com/hashicorp/packer/command" in any of:
	/home/users/bhundven/projects/go/src/github.com/hashicorp/packer/command (from $GOROOT)
	/root/go/src/github.com/hashicorp/packer/command (from $GOPATH)
checkpoint.go:10:2: cannot find package "github.com/hashicorp/packer/packer" in any of:
	/home/users/bhundven/projects/go/src/github.com/hashicorp/packer/packer (from $GOROOT)
	/root/go/src/github.com/hashicorp/packer/packer (from $GOPATH)
config.go:16:2: cannot find package "github.com/hashicorp/packer/packer/plugin" in any of:
	/home/users/bhundven/projects/go/src/github.com/hashicorp/packer/packer/plugin (from $GOROOT)
	/root/go/src/github.com/hashicorp/packer/packer/plugin (from $GOPATH)
checkpoint.go:11:2: cannot find package "github.com/hashicorp/packer/version" in any of:
	/home/users/bhundven/projects/go/src/github.com/hashicorp/packer/version (from $GOROOT)
	/root/go/src/github.com/hashicorp/packer/version (from $GOPATH)
config.go:17:2: cannot find package "github.com/kardianos/osext" in any of:
	/home/users/bhundven/projects/go/src/github.com/kardianos/osext (from $GOROOT)
	/root/go/src/github.com/kardianos/osext (from $GOPATH)
commands.go:6:2: cannot find package "github.com/mitchellh/cli" in any of:
	/home/users/bhundven/projects/go/src/github.com/mitchellh/cli (from $GOROOT)
	/root/go/src/github.com/mitchellh/cli (from $GOPATH)
main.go:24:2: cannot find package "github.com/mitchellh/panicwrap" in any of:
	/home/users/bhundven/projects/go/src/github.com/mitchellh/panicwrap (from $GOROOT)
	/root/go/src/github.com/mitchellh/panicwrap (from $GOPATH)
main.go:25:2: cannot find package "github.com/mitchellh/prefixedio" in any of:
	/home/users/bhundven/projects/go/src/github.com/mitchellh/prefixedio (from $GOROOT)
	/root/go/src/github.com/mitchellh/prefixedio (from $GOPATH)

any idea?

when I look at hashicorp/packer/post-processor/vsphere-template/step_mark_as_template.go line 48 is what you posted above, not whats in the current branch from hashicorp/packer. Then, building the new packer binary with go build I still get this error when making it as a template.

panic: runtime error: index out of range
2017/11/07 15:01:25 packer:
2017/11/07 15:01:25 packer: goroutine 109 [running]:
2017/11/07 15:01:25 packer: github.com/hashicorp/packer/post-processor/vsphere-template.(*stepMarkAsTemplate).Run(0xc42042ce60, 0x30ded60, 0xc420011440, 0x0)
2017/11/07 15:01:25 packer: 	/root/go/src/github.com/hashicorp/packer/post-processor/vsphere-template/step_mark_as_template.go:48 +0x9a9
2017/11/07 15:01:25 packer: github.com/hashicorp/packer/vendor/github.com/mitchellh/multistep.(*BasicRunner).Run(0xc4203e59c0, 0x30ded60, 0xc420011440)
2017/11/07 15:01:25 packer: 	/root/go/src/github.com/hashicorp/packer/vendor/github.com/mitchellh/multistep/basic_runner.go:70 +0x1ee
2017/11/07 15:01:25 packer: github.com/hashicorp/packer/post-processor/vsphere-template.(*PostProcessor).PostProcess(0xc42014e300, 0x30e1ee0, 0xc42042c220, 0x30e3da0, 0xc42042c240, 0x0, 0x0, 0xc420011e00, 0x0, 0x0)
2017/11/07 15:01:25 packer: 	/root/go/src/github.com/hashicorp/packer/post-processor/vsphere-template/post-processor.go:126 +0x5ef
2017/11/07 15:01:25 packer: github.com/hashicorp/packer/packer/rpc.(*PostProcessorServer).PostProcess(0xc4202c8960, 0x1, 0xc420045ef0, 0x0, 0x0)
2017/11/07 15:01:25 packer: 	/root/go/src/github.com/hashicorp/packer/packer/rpc/post_processor.go:83 +0x1a8
2017/11/07 15:01:25 packer: reflect.Value.call(0xc420159320, 0xc4201422e0, 0x13, 0x20551d1, 0x4, 0xc4203bdf20, 0x3, 0x3, 0x0, 0x0, ...)
2017/11/07 15:01:25 packer: 	/home/users/bhundven/projects/go/src/reflect/value.go:434 +0x906
2017/11/07 15:01:25 packer: reflect.Value.Call(0xc420159320, 0xc4201422e0, 0x13, 0xc42002d720, 0x3, 0x3, 0x0, 0x0, 0xc420242ae0)
2017/11/07 15:01:25 packer: 	/home/users/bhundven/projects/go/src/reflect/value.go:302 +0xa4
2017/11/07 15:01:25 packer: net/rpc.(*service).call(0xc420139300, 0xc42013b2c0, 0xc420016f28, 0xc42014b180, 0xc4202c8b80, 0x1ad4960, 0xc420017100, 0x18a, 0x1a0c6a0, 0xc420045ef0, ...)
2017/11/07 15:01:25 packer: 	/home/users/bhundven/projects/go/src/net/rpc/server.go:381 +0x142
2017/11/07 15:01:25 packer: created by net/rpc.(*Server).ServeCodec
2017/11/07 15:01:25 packer: 	/home/users/bhundven/projects/go/src/net/rpc/server.go:475 +0x36b
2017/11/07 15:01:25 Deleting original artifact for build 'CentOS-7-x64'
2017/11/07 15:01:25 /usr/local/bin/packer: plugin process exited

any idea?

--- a/builder/vmware/iso/artifact.go
+++ b/builder/vmware/iso/artifact.go
@@ -4,6 +4,12 @@ import (
 	"fmt"
 )
 
+const (
+	ArtifactConfFormat         = "artifact.conf.format"
+	ArtifactConfKeepRegistered = "artifact.conf.keep_registered"
+	ArtifactConfSkipExport     = "artifact.conf.skip_export"
+)
+
 // Artifact is the result of running the VMware builder, namely a set
 // of files associated with the resulting machine.
 type Artifact struct {
@@ -11,6 +17,7 @@ type Artifact struct {
 	id        string
 	dir       OutputDir
 	f         []string
+	config    map[string]string
 }
 
 func (a *Artifact) BuilderId() string {
@@ -30,7 +37,7 @@ func (a *Artifact) String() string {
 }
 
 func (a *Artifact) State(name string) interface{} {
-	return nil
+	return a.config[name]
 }
 
 func (a *Artifact) Destroy() error {

--- a/builder/vmware/iso/builder.go
+++ b/builder/vmware/iso/builder.go
@@ -6,6 +6,7 @@ import (
 	"io/ioutil"
 	"log"
 	"os"
+	"strconv"
 	"time"
 
 	vmwcommon "github.com/hashicorp/packer/builder/vmware/common"
@@ -343,7 +344,7 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
 
 	// Compile the artifact list
 	var files []string
-	if b.config.RemoteType != "" && b.config.Format != "" {
+	if b.config.RemoteType != "" && b.config.Format != "" && !b.config.SkipExport {
 		dir = new(vmwcommon.LocalOutputDir)
 		dir.SetOutputDir(exportOutputPath)
 		files, err = dir.ListFiles()
@@ -360,11 +361,17 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
 		builderId = BuilderIdESX
 	}
 
+	config := make(map[string]string)
+	config[ArtifactConfKeepRegistered] = strconv.FormatBool(b.config.KeepRegistered)
+	config[ArtifactConfFormat] = b.config.Format
+	config[ArtifactConfSkipExport] = strconv.FormatBool(b.config.SkipExport)
+
 	return &Artifact{
 		builderId: builderId,
 		id:        b.config.VMName,
 		dir:       dir,
 		f:         files,
+		config:    config,
 	}, nil
 }
 

--- a/builder/vmware/iso/step_register.go
+++ b/builder/vmware/iso/step_register.go
@@ -51,7 +51,7 @@ func (s *StepRegister) Cleanup(state multistep.StateBag) {
 	}
 
 	if remoteDriver, ok := driver.(RemoteDriver); ok {
-		if s.Format == "" {
+		if s.Format == "" || config.SkipExport {
 			ui.Say("Unregistering virtual machine...")
 			if err := remoteDriver.Unregister(s.registeredPath); err != nil {
 				ui.Error(fmt.Sprintf("Error unregistering VM: %s", err))

--- a/post-processor/vsphere-template/post-processor.go
+++ b/post-processor/vsphere-template/post-processor.go
@@ -88,13 +88,6 @@ func (p *PostProcessor) PostProcess(ui packer.Ui, artifact packer.Artifact) (pac
 		return nil, false, fmt.Errorf("Unknown artifact type, can't build box: %s", artifact.BuilderId())
 	}
 
-	source := ""
-	for _, path := range artifact.Files() {
-		if strings.HasSuffix(path, ".vmx") {
-			source = path
-			break
-		}
-	}
 	// In some occasions the VM state is powered on and if we immediately try to mark as template
 	// (after the ESXi creates it) it will fail. If vSphere is given a few seconds this behavior doesn't reappear.
 	ui.Message("Waiting 10s for VMware vSphere to start")
@@ -119,12 +112,10 @@ func (p *PostProcessor) PostProcess(ui packer.Ui, artifact packer.Artifact) (pac
 		},
 		&stepMarkAsTemplate{
 			VMName: artifact.Id(),
-			Source: source,
 		},
 	}
 	runner := common.NewRunnerWithPauseFn(steps, p.config.PackerConfig, ui, state)
 	runner.Run(state)
-
 	if rawErr, ok := state.GetOk("error"); ok {
 		return nil, false, rawErr.(error)
 	}

--- a/post-processor/vsphere-template/post-processor.go
+++ b/post-processor/vsphere-template/post-processor.go
@@ -2,11 +2,13 @@ package vsphere_template
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"net/url"
 	"strings"
 	"time"
 
+	"github.com/hashicorp/packer/builder/vmware/iso"
 	"github.com/hashicorp/packer/common"
 	"github.com/hashicorp/packer/helper/config"
 	"github.com/hashicorp/packer/packer"
@@ -86,6 +88,14 @@ func (p *PostProcessor) Configure(raws ...interface{}) error {
 func (p *PostProcessor) PostProcess(ui packer.Ui, artifact packer.Artifact) (packer.Artifact, bool, error) {
 	if _, ok := builtins[artifact.BuilderId()]; !ok {
 		return nil, false, fmt.Errorf("Unknown artifact type, can't build box: %s", artifact.BuilderId())
+	}
+
+	f := artifact.State(iso.ArtifactConfFormat)
+	k := artifact.State(iso.ArtifactConfKeepRegistered)
+	s := artifact.State(iso.ArtifactConfSkipExport)
+
+	if f != "" && k != "true" && s == "false" {
+		return nil, false, errors.New("To use this post-processor with exporting behavior you need set keep_registered as true")
 	}
 
 	// In some occasions the VM state is powered on and if we immediately try to mark as template

--- a/post-processor/vsphere-template/step_mark_as_template.go
+++ b/post-processor/vsphere-template/step_mark_as_template.go
@@ -39,7 +39,7 @@ func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction 
 		return multistep.ActionHalt
 	}
 
-	path, err := datastorePath(vm, s.VMName)
+	dsPath, err := datastorePath(vm)
 	if err != nil {
 		state.Put("error", err)
 		ui.Error(err.Error())
@@ -59,7 +59,7 @@ func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction 
 		return multistep.ActionHalt
 	}
 
-	task, err := folder.RegisterVM(context.Background(), path.String(), s.VMName, true, nil, host)
+	task, err := folder.RegisterVM(context.Background(), dsPath.String(), s.VMName, true, nil, host)
 	if err != nil {
 		state.Put("error", err)
 		ui.Error(err.Error())
@@ -75,7 +75,7 @@ func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction 
 	return multistep.ActionContinue
 }
 
-func datastorePath(vm *object.VirtualMachine, name string) (*object.DatastorePath, error) {
+func datastorePath(vm *object.VirtualMachine) (*object.DatastorePath, error) {
 	devices, err := vm.Device(context.Background())
 	if err != nil {
 		return nil, err
@@ -92,15 +92,15 @@ func datastorePath(vm *object.VirtualMachine, name string) (*object.DatastorePat
 	}
 
 	if disk == "" {
-		return nil, fmt.Errorf("disk not found in '%v'", name)
+		return nil, fmt.Errorf("disk not found in '%v'", vm.Name())
 	}
 
 	re := regexp.MustCompile("\\[(.*?)\\]")
 
 	datastore := re.FindStringSubmatch(disk)[1]
-	vmx := path.Join("/", path.Dir(strings.Split(disk, " ")[1]), name+".vmx")
+	vmxPath := path.Join("/", path.Dir(strings.Split(disk, " ")[1]), vm.Name()+".vmx")
 
-	return &object.DatastorePath{datastore, vmx}, nil
+	return &object.DatastorePath{datastore, vmxPath}, nil
 }
 
 // We will use the virtual machine created by vmware-iso builder
@@ -117,7 +117,12 @@ func findRuntimeVM(cli *govmomi.Client, dcPath, name string) (*object.VirtualMac
 		return nil, fmt.Errorf("VM at path %s not found", fullPath)
 	}
 
-	return ref.(*object.VirtualMachine), nil
+	vm := ref.(*object.VirtualMachine)
+	if vm.InventoryPath == "" {
+		vm.SetInventoryPath(fullPath)
+	}
+
+	return vm, nil
 }
 
 // If in the target folder a virtual machine or template already exists

--- a/post-processor/vsphere-template/step_mark_as_template.go
+++ b/post-processor/vsphere-template/step_mark_as_template.go
@@ -4,17 +4,18 @@ import (
 	"context"
 	"fmt"
 	"path"
+	"regexp"
 	"strings"
 
 	"github.com/hashicorp/packer/packer"
 	"github.com/mitchellh/multistep"
 	"github.com/vmware/govmomi"
 	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
 )
 
 type stepMarkAsTemplate struct {
 	VMName string
-	Source string
 }
 
 func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction {
@@ -26,6 +27,19 @@ func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction 
 	ui.Message("Marking as a template...")
 
 	vm, err := findRuntimeVM(cli, dcPath, s.VMName)
+	if err != nil {
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	if err := unregisterPreviousVM(cli, folder, s.VMName); err != nil {
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	path, err := datastorePath(vm, s.VMName)
 	if err != nil {
 		state.Put("error", err)
 		ui.Error(err.Error())
@@ -45,21 +59,7 @@ func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction 
 		return multistep.ActionHalt
 	}
 
-	source := strings.Split(s.Source, "/vmfs/volumes/")[1]
-	i := strings.Index(source, "/")
-
-	path := (&object.DatastorePath{
-		Datastore: source[:i],
-		Path:      source[i:],
-	}).String()
-
-	if err := unregisterPreviousVM(cli, folder, s.VMName); err != nil {
-		state.Put("error", err)
-		ui.Error(err.Error())
-		return multistep.ActionHalt
-	}
-
-	task, err := folder.RegisterVM(context.Background(), path, s.VMName, true, nil, host)
+	task, err := folder.RegisterVM(context.Background(), path.String(), s.VMName, true, nil, host)
 	if err != nil {
 		state.Put("error", err)
 		ui.Error(err.Error())
@@ -73,6 +73,34 @@ func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction 
 	}
 
 	return multistep.ActionContinue
+}
+
+func datastorePath(vm *object.VirtualMachine, name string) (*object.DatastorePath, error) {
+	devices, err := vm.Device(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	disk := ""
+	for _, device := range devices {
+		if d, ok := device.(*types.VirtualDisk); ok {
+			if b, ok := d.Backing.(types.BaseVirtualDeviceFileBackingInfo); ok {
+				disk = b.GetVirtualDeviceFileBackingInfo().FileName
+			}
+			break
+		}
+	}
+
+	if disk == "" {
+		return nil, fmt.Errorf("disk not found in '%v'", name)
+	}
+
+	re := regexp.MustCompile("\\[(.*?)\\]")
+
+	datastore := re.FindStringSubmatch(disk)[1]
+	vmx := path.Join("/", path.Dir(strings.Split(disk, " ")[1]), name+".vmx")
+
+	return &object.DatastorePath{datastore, vmx}, nil
 }
 
 // We will use the virtual machine created by vmware-iso builder

